compiler/qsc_linter/src/lints/hir.rs

```diff
@@ -1,17 +1,106 @@
+struct IsQuantumOperation {
+    /// This field is set to `true` after calling `Visitor::visit_callable_decl(...)`
+    /// if the operation satisfies the characteristics described above.
+    is_op: bool,
 }
 
-impl HirLintPass for Placeholder {
-    fn check_expr(&self, expr: &qsc_hir::hir::Expr, buffer: &mut Vec<Lint>) {
-        if let qsc_hir::hir::ExprKind::Lit(Lit::Int(42)) = &expr.kind {
-            buffer.push(lint!(self, expr.span));
+impl IsQuantumOperation {
+    /// Returns `true` if the declaration is empty.
+    fn is_empty_decl(spec_decl: Option<&SpecDecl>) -> bool {
+        match spec_decl {
+            None => true,
+            Some(decl) => match &decl.body {
+                SpecBody::Gen(_) => true,
+                SpecBody::Impl(_, block) => block.stmts.is_empty(),
+            },
+        }
+    }
+
+    /// Returns `true` if the operation is empty.
+    /// An operation is empty if there is no code for its body and specializations.
+    fn is_empty_op(call_decl: &CallableDecl) -> bool {
+        Self::is_empty_decl(Some(&call_decl.body))
+            && Self::is_empty_decl(call_decl.adj.as_ref())
+            && Self::is_empty_decl(call_decl.ctl.as_ref())
+            && Self::is_empty_decl(call_decl.ctl_adj.as_ref())
+    }
+}
+
+impl Visitor<'_> for IsQuantumOperation {
+    fn visit_callable_decl(&mut self, decl: &CallableDecl) {
+        if Self::is_empty_op(decl) {
+            self.is_op = true;
+        } else {
+            visit::walk_callable_decl(self, decl);
+        }
+    }
+
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        if !self.is_op {
+            if let StmtKind::Qubit(..) = &stmt.kind {
+                self.is_op = true;
+            } else {
+                visit::walk_stmt(self, stmt);
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        if !self.is_op {
+            match &expr.kind {
+                ExprKind::Call(callee, _) => {
+                    if matches!(&callee.ty, Ty::Arrow(arrow) if arrow.kind == CallableKind::Operation)
+                    {
+                        self.is_op = true;
+                    }
+                }
+                ExprKind::Conjugate(..) | ExprKind::Repeat(..) => {
+                    self.is_op = true;
+                }
+                _ => {
+                    visit::walk_expr(self, expr);
+                }
+            }
+        }
+    }
+}
+
+/// HIR Lint for [`NeedlessOperation`], suggesting to use function
+/// We use [`IsQuantumOperation`] helper to check if a operation has desired operation characteristics
+impl HirLintPass for NeedlessOperation {
+    fn check_callable_decl(&self, decl: &CallableDecl, buffer: &mut Vec<Lint>) {
+        if decl.kind == CallableKind::Operation {
+            let mut op_limits = IsQuantumOperation::default();
+
+            op_limits.visit_callable_decl(decl);
+
+            if !op_limits.is_op {
+                buffer.push(lint!(self, decl.span));
```

Comment:
Just to make the squiggles a little less obtrusive while I'm still typing up the operation, can we use `decl.name.span` for this? 
```suggestion
                buffer.push(lint!(self, decl.name.span));
```

![image](https://github.com/microsoft/qsharp/assets/16928427/1480d496-d74b-477b-bd68-7208782b8dea)